plugins/steelseries/fu-steelseries-fizz-tunnel.c

```diff
@@ -0,0 +1,407 @@
+
+		/* success */
+		return TRUE;
+	}
+	fu_device_set_version(device, version);
+
+	if (!fu_steelseries_fizz_file_crc32(parent,
+					    TRUE,
+					    fs,
+					    id,
+					    &calculated_crc,
+					    &stored_crc,
+					    &error_local)) {
+		g_debug("ignoring error on get file CRC32 from FS 0x%02x ID 0x%02x: %s", fs, id, error_local->message);
+
+		/* success */
+		return TRUE;
+	}
+
+	if (calculated_crc != stored_crc)
+		g_warning("%s: checksum mismatch, got 0x%08x, expected 0x%08x",
+			  fu_device_get_name(device),
+			  calculated_crc,
+			  stored_crc);
+
+	fu_device_remove_flag(device, FWUPD_DEVICE_FLAG_UNREACHABLE);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_fizz_tunnel_write_firmware(FuDevice *device,
+					  FuFirmware *firmware,
+					  FuProgress *progress,
+					  FwupdInstallFlags flags,
+					  GError **error)
+{
+	FuDevice *parent = fu_device_get_parent(device);
+	guint8 fs = STEELSERIES_FIZZ_FILESYSTEM_MOUSE;
+	guint8 id = STEELSERIES_FIZZ_MOUSE_FILESYSTEM_BACKUP_APP_ID;
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 100);
+
+	if (!fu_steelseries_fizz_write_file(parent,
+					    TRUE,
+					    fs,
+					    id,
+					    firmware,
+					    fu_progress_get_child(progress),
+					    flags,
+					    error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_steelseries_fizz_tunnel_read_firmware(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent(device);
+	guint8 fs = STEELSERIES_FIZZ_FILESYSTEM_MOUSE;
+	guint8 id = STEELSERIES_FIZZ_MOUSE_FILESYSTEM_BACKUP_APP_ID;
+	gsize bufsz = 0x27000;
+	g_autoptr(FuFirmware) firmware = fu_steelseries_firmware_new();
+	g_autoptr(GBytes) blob = NULL;
+	g_autofree guint8 *buf = NULL;
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_READ, 100);
+
+	buf = g_malloc0(bufsz);
+	if (!fu_steelseries_fizz_read_access_file(parent,
+						  TRUE,
+						  fs,
+						  id,
+						  buf,
+						  bufsz,
+						  fu_progress_get_child(progress),
+						  error))
+		return NULL;
+	fu_progress_step_done(progress);
+
+	if (g_getenv("FWUPD_STEELSERIES_FIZZ_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "Firmware", buf, bufsz);
+	blob = g_bytes_new_take(g_steal_pointer(&buf), bufsz);
+	if (!fu_firmware_parse(firmware, blob, FWUPD_INSTALL_FLAG_NONE, error))
+		return NULL;
+
+	/* success */
+	return g_steal_pointer(&firmware);
+}
+
+static FuFirmware *
+fu_steelseries_fizz_tunnel_prepare_firmware(FuDevice *device,
+					    GBytes *fw,
+					    FwupdInstallFlags flags,
+					    GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_steelseries_firmware_new();
+
+	if (!fu_firmware_parse(FU_FIRMWARE(firmware), fw, flags, error))
+		return NULL;
+
+	/* success */
+	return g_steal_pointer(&firmware);
+}
```

Comment:
if do you fu_device_set_firmware_gtype(device, FU_TYPE_STEELSERIES_FIRMWARE) then you don't need the prepare_firmware section at all.

Reply:
Also, is the firmware specific to Fizz? in which case it should probably be FU_TYPE_STEELSERIES_FIZZ_FIRMWARE -- otherwise it's fine.

Reply:
The firmware (blob) is common to fizz and Sonic. It is a blob with trailing crc32. I wonder if gamepad is the same type.

Reply:
Sonic is using an archive of three crc32-firmwares.

Reply:
So maybe FU_TYPE_STEELSERIES_CHECKSUM_FIRMWARE? or as it is now is fine too. Maybe KISS and leave it as you've got.

Reply:
it makes sense, tbh I wondered if some other devices use a CRC at the end of the blob; if so it could be a `FU_TYPE_CHECKSUM_FIRMWARE` with a bit more intelligent handling the type or checksum.

Reply:
Stick to what you have now, and we can merge/superclass the plugin FuFirmware in the future. My normal guideline is 3, i.e. if three plugins are doing the same thing then we abstract it out. We have 2 now.

Reply:
I assumes, I can save code from [fizz](https://github.com/fwupd/fwupd/blob/main/plugins/steelseries/fu-steelseries-fizz.c#L704-L717) as well

Reply:
[c8a4086](https://github.com/fwupd/fwupd/pull/4670/commits/c8a4086b474d6cf8e91c8d2795e2359ba736b202)